cassandane/Cassandane/Instance.pm

```diff
@@ -1426,24 +1432,29 @@ sub _detect_core_program
     return $prog;
 }
 
-sub _check_cores
+sub find_cores
 {
     my ($self) = @_;
-
     my $coredir = $self->{basedir} . '/conf/cores';
-    my $ncores = 0;
+
+    my $cassini = Cassandane::Cassini->instance();
+    my $core_pattern = $cassini->val('cassandane', 'core_pattern',
+                                     '^core.*?(?:\.\d+)?$');
```

Comment:
This pattern is significantly different than the old one. Before we'd strictly match core.1234 (period here being the actual '.' character). Now we'll match anything that starts with the word 'core'.
Since (?:\.\d+) is followed by ?, it can match 0 times, which means the pattern is really just /^core.*$/

Reply:
This is the detail that I most wanted your feedback on, so thanks :)

We used to match either exactly "core" or "core.%p".  This all came up because Fastmail's test system has "core.%E.%t.%s.%p", so my goal was to also be able to match that in the default pattern, without getting too specific about the inner fields.

I thought about looking specifically for dot-separated fields, but while I was thinking about that it occurred to me that if I were configuring my own system, I would probably use hyphens for the non-pid fields, so I ought to match those too, and then I decided to not sweat the details too much and just match anything between "core" and ".%p".  But you're right, with the ".%p" being optional, it becomes redundant because the `.*?` would match it anyway.

Right now, I'm torn between simplifying it to just `^core.*$` (which it effectively is already), or changing it to something like `^core(?:.*?\.\d+)?$` (which maybe better captures my original intent, but reduces its generality).  What do you think?

Reply:
I also keep going both ways on whether that `.*?` needs to be non-greedy or if `.*` would be fine.  It's a relatively simple big-picture problem so I'm getting stuck on the overthinking the details 😅

Reply:
I think there's probably three goals
 * Don't change behaviour and cause breakage where there was none before
 * Don't increase false positives
 * Don't increase false negatives

It's hard to know about the latter 2 goals. In the interest of not having to think too hard about it, I wouldn't change the default pattern from the one in master at this point in time, but that's probably just me being lazy :)

Reply:
The other detail I guess is that the Test::Core tests examine the filename of the core file they found, and if it contains a series of digits, we assert that it matches the pid of our crasher.  If someone has a complicated system core format where pid is not the first numeric term, perhaps say "core.timestamp.pid", even if they configure cassandane's core_pattern to detect those files, the Test::Core will mistake the timestamp part of the filename for a pid and then complain that it doesn't match the crashing process (because it's possible that something crashed other than the thing we intended to crash, invalidating the test).

I could change Test::Core to extract _all_ numeric sections from the filename and assert that if any are present, at least one of them matches... but that would still be a problem for core formats that have numeric fields but the pid isn't one of them.

Maybe instead Test::Core should also just apply the configured core_pattern, and _iff there were a capture group that matched_, compare its contents to the pid.  That way someone configuring a custom core_pattern can control if or where the pid field should be found.  And in that case I'd make the default core_pattern something like `/^core.*?(?:\.(\d+))?$/`, intending to still match anything starting with "core" as we do already, but if it ends with .pid then the pid can be captured and checked.  If someone's pid field is present but not at the end, they can configure their core_pattern with the capture group in the appropriate place.

Reply:
Yeah... that's an option. Another would be to use a named capture that the user must use in their core pattern:
/^core.*?(?<pid>\d+)\.pid/

And then you can check `$+{pid}`... but maybe that's more complicated.

Reply:
Yeah I haven't used named captures myself before.  Saw them in the docs while finding the correct vocabulary for "capture group", but if I haven't used them as someone who spends roughly half my programming time writing perl, the chances some sysadmin has are slim!  If we were trying to capture multiple fields that might be in any order, they'd be perfect, but since we're only after one field, I think this is a case of do-the-simpler-thing.

I mean, hopefully most people never need to use the core_pattern setting anyway, especially now that the matching is more general than it initially was.  But if they do need to change it, at this point with the capture group for pid I think they'll be able to do what they need without needing further code changes.